Mesos DNS changes in scheduler

abhay-agarwal opened the pull request.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -106,13 +113,53 @@ public void frameworkMessage(ExecutorDriver driver, byte[] msg) {
-        startProcess(driver, nameNodeTask);
-        startProcess(driver, zkfcNodeTask);
+      }
+      runCommand(driver, nameNodeTask, "bin/hdfs-mesos-namenode " + messageStr);
```
I think this runCommand(init/bootstrap) does belong up in the else case, since we don't want to reformat the NN metadata if it already exists. The rest of the startProcess(NN + ZKFC) and sendStatusUpdate(RUNNING) lines probably belong outside of the if/else as you have them.

elingg replied:
Right, we don't want to reformat every time. Everything should be within the else. We start the nodes when the task is launched. We only want to restart them in the case of reformatting.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -65,6 +69,9 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo taskInfo) {
           .build());
+      Timer timer = new Timer();
+      PreNNInitTask initTask = new PreNNInitTask(driver, nameNodeTask);
+      timer.scheduleAtFixedRate(initTask, 30000, 15000);
```
Should only start this process/timer if usingMesosDNS, right?

adam-mesos commented on conf/mesos-site.xml:
```diff
@@ -53,6 +53,16 @@
   </property>
 
   <property>
+    <name>mesos.hdfs.mesosdns</name>
```
Why have the boolean instead of just defaulting 'mesos.hdfs.mesosdns.domain' to EmptyString, and then use if (conf.getMesosDnsDomain() != "") instead of conf.usingMesosDns()?

abhay-agarwal replied:
It's for usability. I originally had it the other way around, but having an empty string denote functionality is really confusing (and this way follows hadoop convention)

adam-mesos replied:
No problem. Let's follow hadoop convention then. I get used to the terseness of Mesos/C++, so sometimes it's hard for me to context-switch back into the verboseness of Java/Hadoop. :)

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -16,17 +16,12 @@
 import java.sql.Timestamp;
-import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.*;
```
I usually prefer not to use 'import *', so that we're clear about specifically what we're importing and there's less of a chance of an internal classname conflicting with an (unused) imported class.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -572,21 +602,57 @@ public void run() {
-          }
-        }
-        if (!taskFound) {
+        if (!runningTaskIds.contains(taskId))
```
Love the change. So much simpler. Now we still need to handle running tasks in liveState that don't exist in persistentState's taskIds. TODO?

elingg replied:
+1 Nice

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -572,21 +602,57 @@ public void run() {
+
+    boolean success = true;
+    for (String host : hosts) {
+      log.info("Checking for " + host);
```
s/Checking for/Resolving IP for/ ?

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -572,21 +602,57 @@ public void run() {
+      log.info("Checking for " + host);
+      try {
+        InetAddress.getByName(host);
+        log.info("Successfully found " + host);
```
Maybe log the actual InetAddress found?

abhay-agarwal replied:
Doesn't matter. The Inet found is just a mesos dns entry, and is already available via persistentstate.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -572,21 +602,57 @@ public void run() {
+  private boolean journalNodesAvailable() {
+    Set<String> hosts = new HashSet<>();
+    if (conf.usingMesosDns()) {
+      for (int i = conf.getJournalNodeCount(); i > 0; i--)
```
Why backwards? To fail earlier if JN-last is unavailable? Worth a comment so others don't naively "fix" your reverse iteration.

abhay-agarwal replied:
it's an easier way to implement journalnode{1-3}, which maintains consistency with hdfs convention.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -572,21 +602,57 @@ public void run() {
     }
   }
+
+  private boolean journalNodesAvailable() {
```
journalNodesResolvable, or journalNodesReachable?

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/state/AcquisitionPhase.java:
```diff
@@ -3,7 +3,9 @@
 public enum AcquisitionPhase {
   RECONCILING_TASKS, // Waits here for the timeout on (re)registration
   JOURNAL_NODES, // Launches and waits for all journalnodes to start
+  WAIT_FOR_JOURNAL_NODES, // waits for journal nodes to be available via DNS
```
Please capitalize the beginning of each comment.

adam-mesos replied:
s/available/resolvable/?

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -21,8 +21,9 @@
   public static final Log log = LogFactory.getLog(LiveState.class);
   private Set<Protos.TaskID> stagingTasks = new HashSet<>();
   private AcquisitionPhase currentAcquisitionPhase = AcquisitionPhase.RECONCILING_TASKS;
-  // TODO (nicgrayson) Might need to split this out to jns, nns, and dns if dns too big
```
Since you're not fixing getJournalNodeSize()/getNameNodeSize() to return the number of running JNs/NNs without iterating (filtering) through all runningTasks, I don't think you can really remove this TODO. Your new HashMaps cover staging and running tasks, so they won't be able to do that job. We may still want to have other separate lists of /running/ JNs and NNs.

elingg commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -250,19 +260,38 @@ public void run() {
     List<TaskInfo> tasks = new ArrayList<>();
-    for (String taskName : taskNames) {
-      List<Resource> taskResources = getTaskResources(taskName);
+    for (String taskType : taskTypes) {
```
Let's break this into a separate method, please since the method is getting large.

elingg commented on src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -106,13 +113,53 @@ public void frameworkMessage(ExecutorDriver driver, byte[] msg) {
+   * This class is designed to help launch the secondary namenode in its bootstrap phase. The node
+   * needs both name nodes to be available as DNS entries. In Mesos-DNS, this takes an uncertain
+   * amount of time, so we must schedule it asynchronously.
+   */
```
You are already doing the check in the scheduler so there's no reason to have this code here. You will not get to the bootstrap code until you are in the proper phase.

abhay-agarwal replied:
Unfortunately, this code is necessary. Since the scheduler only formats the nodes in response to a status update, the node needs to send a status update when its discoverable. This is equivalent to the status update that gets sent after a node is launched.

elingg replied:
The scheduler could be changed to format the nodes after it finds that they are running, but also it needs to check that they are discoverable which is a call on the scheduler. Why is it that you can't transition states once DNS is resolvable in the scheduler? The other option I see here is to have the methods MesosDns.journalNodesResolvable(conf, persistentState) called from the Executors instead of from the scheduler and use status messages to make the appropriate transitions. Either way will be fine, I just want to make sure we aren't duplicating code between the scheduler and executors unnecessarily. Let me know if I'm missing something, and we can also discuss tomorrow if needed.

elingg commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -572,21 +602,57 @@ public void run() {
       }
       reconciliationCompleted = true;
     }
   }
```
The scheduler is getting really bloated. Can you separate these methods into a separate class. Also, can you add some kind of timeout, perhaps for 30 seconds, so this is not called every time there is a resource offer?  private RateLimiter reloadLimiter = RateLimiter.create(1 / 30.); // no more than once every 30s

abhay-agarwal replied:
rewrote this so that it doesn't get executed as often. It only executes it while loading its node type.

abhay-agarwal commented:
Please check the new changes @elingg @adam-mesos.

I moved the DNS resolution code into a separate file and cleaned up the phases. Instead of adding extra phases, I decided to move the logic into the state machine because it simplifies the state transitions. Took out the logic in the executor and turned that into a `sendMessageAfterNNResolvable` method.

I'm confident about the functionality here and its ready to merge IMHO.

abhay-agarwal commented on src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -106,15 +106,15 @@ public void frameworkMessage(ExecutorDriver driver, byte[] msg) {
+          .setTaskId(nameNodeTask.taskInfo.getTaskId())
+          .setState(TaskState.TASK_RUNNING)
+          .setMessage(messageStr)
+          .build());
```
I found that this is the only strategy that works. Adam and I discussed, and maybe there's a way to use a format without the `-force` flag, but unless you format the node, the other name node won't be able to bootstrap.

elingg replied:
This will create a large bug in our HA features so this is not mergable. Basically you are reformatting the NN when it's been already formatted. I'm guessing the issue you are encountering is when you want to redeploy HDFS and start from scratch without recovering data. In that case, I suggest you remove /tmp/hdfs data directories.

adam-mesos replied:
So now the "hdfs-mesos-namenode -i/b" is only in the else case. That's the part that does the reformatting, right?
Seems like if the scheduler sends a frameworkMessage(init/bootstrap) to the node, then it should send back the TASK_RUNNING(init/bootstrap) regardless of whether it reformatted. But real data could convince me otherwise.
The question of whether/when you need to startProcess(NN,ZKFC) is still uncertain to me.

elingg replied:
@abhay-agarwal and I discussed. It needs to restart the tasks and send status running, but not reformat.

abhay-agarwal commented on src/main/java/org/apache/mesos/hdfs/state/PersistentState.java:
```diff
@@ -176,15 +179,15 @@ public void removeTaskId(String taskId) {
 
   public boolean dataNodeRunningOnSlave(String hostname) {
-    return getDataNodes().keySet().contains(hostname);
+    return getDataNodes().containsKey(hostname);
```
I modified these because it seemed like an oversight to derive the keyset first. But let me know if there was some other reasoning there.

elingg replied:
Looks good. This is a code style improvement.

abhay-agarwal commented on src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -46,19 +58,21 @@ public void removeStagingTask(final Protos.TaskID taskID) {
         && nameNode1TaskMap.keySet().iterator().next().getTaskId().equals(taskId)) {
       nameNode1TaskMap.clear();
-    } else if (isNameNode2Initialized()
+    } else if (!nameNode2TaskMap.isEmpty()
```
Depending on the error state, these should be cleared even if they aren't initialized.

elingg replied:
Ah, got it, thanks. Good point.

adam-mesos replied:
+1

elingg commented:
Thanks for your changes @abhay-agarwal! This solution has evolved and is much simpler and cleaner without the duplicated logic. It is looking good. I will test it out today and look it over one last time. My main remaining concern is the bug automatically deleting the Active NN directory. We need to revert that logic.

abhay-agarwal commented:
Okay @elingg, i'm fine with changing the logic around, but I would just like to see in practice. the recovery you describe. Also, the Name node needs to send its status update no matter what, so there was still a bug in your former code, correct?

elingg commented:
Hey @abhay-agarwal, we discussed in person and you made the update that is correct in your last commit. Let me do a final test, look it over one last time, and once @adam-mesos is ready we can merge

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -250,19 +248,21 @@ public void run() {
-      List<Resource> taskResources = getTaskResources(taskName);
+    for (String taskType : taskTypes) {
+      List<Resource> taskResources = getTaskResources(taskType);
+      String taskName = getNodeName(taskType);
```
Suggestion: How about `taskName = getNextTaskName(taskType)`?

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -271,14 +271,35 @@ private void launchNode(SchedulerDriver driver, Offer offer,
+
+  private String getNodeName(String taskType) {
+    if (taskType.equals(HDFSConstants.NAME_NODE_ID)) {
+      for (int i = HDFSConstants.TOTAL_NAME_NODES; i > 0; i--) {
```
Usability/confusability issue? Doesn't this mean that we'll launch a task named NN2 before launching NN1?
Seems like it would be easy enough to say `for (int i = 1; i <= TOTAL_NNs; i++)` to get 1-based naming.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/Scheduler.java:
```diff
@@ -250,19 +248,21 @@ public void run() {
-  private void launchNode(SchedulerDriver driver, Offer offer,
-        String nodeName, List<String> taskNames, String executorName) {
+  private boolean launchNode(SchedulerDriver driver, Offer offer,
+        String nodeName, List<String> taskTypes, String executorName) {
```
Suggestion: Would be great to see a comment somewhere about the differences/relationship between nodeName, taskType, taskIdName, executorName, taskName, and taskId.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -34,8 +36,18 @@ public boolean isNameNode2Initialized() {
+    if (taskId.getValue().contains(HDFSConstants.NAME_NODE_TASKID)) {
+      nameNodeNames.put(taskId.getValue(), taskName);
+    }
+    if (taskId.getValue().contains(HDFSConstants.JOURNAL_NODE_ID)) {
```
Nits: else if? Any reason to check the NNs before the JNs? So far, we've been maintaining a JN,NN,DN ordering to match the launch order.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/util/DnsResolver.java:
```diff
@@ -0,0 +1,96 @@
+
+/**
+ * Created by Abhay Agarwal on 3/12/15.
+ */
```
What's the point of this comment? The git logs can tell us who created what file when. If we end up with a large enough set of contributors that we need an OWNERS file in the future, then you can be the owner of DnsResolver. Until then, let's use class comments for explaining what the class actually does.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/state/PersistentState.java:
```diff
@@ -25,6 +27,7 @@
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
+@Singleton
```
Question: Why make this a Singleton? And if it is now a singleton, do you still need the member variables to be 'static'?

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/util/DnsResolver.java:
```diff
@@ -0,0 +1,96 @@
+    for (String host : hosts) {
+      log.info("Resolving DNS for " + host);
+      try {
+        InetAddress.getByName(host);
```
Can you explain what problem you're trying to solve here by waiting until the NNs/JNs are "resolvable"? I didn't see a github issue attached to this PR that describes the problem. Seems like without mesos-dns, we would just rely on TASK_LOST/FAILED updates rather than worrying about resolving the hosts. If that's the case, why wouldn't we just return true immediately if !usingMesoDns?

elingg replied:
Ah, good point. We should return true immediately if not using Mesos DNS for sure to avoid executing the logic.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/util/DnsResolver.java:
```diff
@@ -0,0 +1,96 @@
+        log.info("Successfully found " + host);
+      } catch (SecurityException | IOException e) {
+        log.info("Couldn't resolve host " + host);
+        log.info("Sleeping before retrying.");
```
Sleeping? Do you mean that the caller (statusUpdate()) will decline the current offer and wait for another offer before trying again to resolve? This log statement belongs up in the caller where the "sleeping" is actually occurring.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/util/DnsResolver.java:
```diff
@@ -0,0 +1,96 @@
+
+  public void sendMessageAfterNNResolvable(final SchedulerDriver driver,
+      final Protos.TaskID taskId,
+      final Protos.SlaveID slaveID, final String message) {
```
Nit: Inconsistent wrapping. Either put each parameter on its own line, or only wrap at line end.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/util/DnsResolver.java:
```diff
@@ -0,0 +1,96 @@
+        log.info("Successfully found " + host);
+      } catch (SecurityException | IOException e) {
+        log.info("Couldn't resolve host " + host);
+        log.info("Sleeping before retrying.");
```
This one's actually delaying until the next 10000ms, but again, this logging belongs in the caller (sendMessageAfterNNResolvable/PreNNInitTask) where the wait is actually happening.

adam-mesos commented on src/main/java/org/apache/mesos/hdfs/util/DnsResolver.java:
```diff
@@ -0,0 +1,96 @@
+  public void sendMessageAfterNNResolvable(final SchedulerDriver driver,
+      final Protos.TaskID taskId,
+      final Protos.SlaveID slaveID, final String message) {
+    class PreNNInitTask extends TimerTask {
```
Rename PreNNInitTask to match the new function/logic, maybe SendMessageIfNNResolvableTask

abhay-agarwal commented:
should be good now, @elingg!

elingg commented:
Awesome! thx. Let me test, take a final look, and then I'll merge.@adam-mesos, do let us know if you have any final objections.

elingg commented:
:100: Let's merge!
